Optimization of DBSCAN with OpenMP

MarkFischinger opened the pull request:
I've been working on optimizing our DBSCAN implementation, and I'm excited to share the results with you! This pull request introduces several performance enhancements that should make the clustering algorithm much faster, especially for larger datasets.

## Changes made

1. Implemented parallel processing for centroids calculation
2. Improved parallelization of cluster assignments
3. Optimized batch clustering with better OpenMP usage
4. Introduced atomic operations to reduce critical sections
5. Enhanced memory locality for better cache utilization


I ran some benchmarks to compare the performance of the original implementation with the optimized version. 


![output](https://github.com/user-attachments/assets/7159566e-7343-4ec7-be1c-d20ee0fb735d)

I've tested this on a variety of datasets, and it seems to be working well.
However, I'd really appreciate if you could take a look and let me know if there's anything I've missed or if you have any suggestions for further improvements.

shrit commented on src/mlpack/methods/dbscan/dbscan_impl.hpp:
```diff
@@ -13,6 +13,9 @@
 #define MLPACK_METHODS_DBSCAN_DBSCAN_IMPL_HPP
 
 #include "dbscan.hpp"
+#include <omp.h>
+#include <vector>
+#include <atomic>
```
These are already included, search for them inside `core.hpp` or `prereqs.hpp`, if they are inside then please include these files,

I also think that you can do the include inside the `dbscan.hpp`, no need to include anything in `impl.hpp`

shrit commented on src/mlpack/methods/dbscan/dbscan_impl.hpp:
```diff
@@ -107,14 +129,25 @@ size_t DBSCAN<RangeSearchType, PointSelectionPolicy>::Cluster(
 
   // Get a count of all clusters.
-  const size_t numClusters = max(assignments) + 1;
+  const size_t numClusters = arma::max(assignments) + 1;
```
any reason to add the `arma`, we removed this intentionally, so if there is no error could you remove it please?

shrit reviewed (commented):
Nice addition, I left some comments

geekypathak21 commented on src/mlpack/methods/dbscan/dbscan_impl.hpp:
```diff
@@ -68,25 +71,44 @@ size_t DBSCAN<RangeSearchType, PointSelectionPolicy>::Cluster(
-  arma::Row<size_t> counts;
-  counts.zeros(numClusters);
-  for (size_t i = 0; i < data.n_cols; ++i)
+  arma::Row<size_t> counts(numClusters, arma::fill::zeros);
```
We are getting rid of `arma::fill::zeros` can you try removing it no worries if you have any specific reason for adding this.

shrit replied:
actually `arma::fill::zeros` is not needed, since the min version of armadillo will do zero initialization by default if we do not have anything specified

geekypathak21 replied:
Oh cool we can remove this for sure then :+1:

shrit replied:
@MarkFischinger  can you remove this? as described in the above comment ? thanks

geekypathak21 commented on src/mlpack/methods/dbscan/dbscan_impl.hpp:
```diff
@@ -68,25 +71,44 @@ size_t DBSCAN<RangeSearchType, PointSelectionPolicy>::Cluster(
+  arma::Row<size_t> counts(numClusters, arma::fill::zeros);
+
+  const size_t numThreads = omp_get_max_threads();
+  std::vector<MatType> localCentroids(numThreads, MatType(data.n_rows, numClusters, arma::fill::zeros));
+  std::vector<arma::Row<size_t>> localCounts(numThreads, arma::Row<size_t>(numClusters, arma::fill::zeros));
```
Same here.

geekypathak21 reviewed (changes requested):
Hey @MarkFischinger Nice work :+1:

rcurtin commented on src/mlpack/methods/dbscan/dbscan_impl.hpp:
```diff
@@ -78,15 +109,20 @@ size_t DBSCAN<RangeSearchType, PointSelectionPolicy>::Cluster(
+  #endif
   for (size_t i = 0; i < numClusters; ++i)
-    centroids.col(i) /= counts[i];
+    if (counts[i] > 0)
```
Why remove the comment pointing out that we are guaranteed the number of points in a cluster is greater than 0?  (Then this check is not needed.)

rcurtin commented on src/mlpack/methods/dbscan/dbscan_impl.hpp:
```diff
@@ -78,15 +109,20 @@ size_t DBSCAN<RangeSearchType, PointSelectionPolicy>::Cluster(
   return numClusters;
 }
 
+
```
```suggestion
```

rcurtin commented on src/mlpack/methods/dbscan/dbscan_impl.hpp:
```diff
@@ -107,14 +143,32 @@ size_t DBSCAN<RangeSearchType, PointSelectionPolicy>::Cluster(
+  #pragma omp parallel for
+  #endif
   for (size_t i = 0; i < data.n_cols; ++i)
     assignments[i] = uf.Find(i);
```
I don't think `UnionFind::Find()` is thread-safe, so I'm not sure that this can work correctly.  Or did I overlook something?

rcurtin reviewed (commented):
I haven't reviewed the whole PR yet, but the initial idea looks good.  There are some comments about `MLPACK_USE_OPENMP` and other things that I left in #3762; do you think you can apply the relevant comments here too?

shrit commented:
@MarkFischinger could you resolve the conflict with the master branch ? thank you

rcurtin commented on src/mlpack/methods/dbscan/dbscan_impl.hpp:
```diff
@@ -68,19 +68,31 @@ size_t DBSCAN<RangeSearchType, PointSelectionPolicy>::Cluster(
 
-  // We should be guaranteed that the number of clusters is always greater than
-  // zero.
+  // Normalize centroids
```
Can you revert to or include the intent of the original comment? Someone wrote that there for a reason.

rcurtin commented on src/mlpack/methods/dbscan/dbscan_impl.hpp:
```diff
@@ -164,53 +184,62 @@ void DBSCAN<RangeSearchType, PointSelectionPolicy>::PointwiseCluster(
 
   for (size_t i = 0; i < data.n_cols; ++i)
   {
-    if (i % 10000 == 0 && i > 0)
```
I believe that all of the optimizations below here are invalid and cause this algorithm to be something other than DBSCAN. I hate to say that because these give nice speedups (and are the core of the speedup of this PR) but inhere tly DBSCAN is a serial algorithm: select a point, find other points within range, mark them as part of the same cluster, repeat.  That can't be parallelized (or I mean you can, but then it's not DBSCAN so we can't do it).

rcurtin commented on src/mlpack/methods/dbscan/dbscan_impl.hpp:
```diff
@@ -164,53 +184,62 @@ void DBSCAN<RangeSearchType, PointSelectionPolicy>::PointwiseCluster(
-      Log::Info << "DBSCAN clustering on point " << i << "..." << std::endl;
-
-    // Get the next index.
-    const size_t index = pointSelector.Select(i, data);
```
The use of `pointSelector` to allow custom next point selection strategies is also important and we shouldn't remove it.

rcurtin commented on src/mlpack/methods/dbscan/dbscan_impl.hpp:
```diff
@@ -164,53 +184,62 @@ void DBSCAN<RangeSearchType, PointSelectionPolicy>::PointwiseCluster(
+      visited[index] = true;
+
+      // Do the range search for only this point.
+      rangeSearch.Search(data.col(index), RangeType<ElemType>(ElemType(0.0), epsilon),
```
Realistically this is the "painful" part of DBSCAN. We would likely get better results by parallelizing the range search itself, but that is a bit of a can of worms, since you would have to parallelize the tree traversal (likely with OpenMP tasks) and then ensure that `RangeSearchRules` is thread-safe.  Not impossible for sure, but probably a good amount of work and tuning.

rcurtin commented on src/mlpack/methods/dbscan/dbscan_impl.hpp:
```diff
@@ -225,7 +254,6 @@ void DBSCAN<RangeSearchType, PointSelectionPolicy>::BatchCluster(
     const MatType& data,
     UnionFind& uf)
 {
-  // For each point, find the points in epsilon-neighborhood and their distances.
```
No need to remove the comment.

rcurtin commented on src/mlpack/methods/dbscan/dbscan_impl.hpp:
```diff
@@ -268,4 +297,4 @@ void DBSCAN<RangeSearchType, PointSelectionPolicy>::BatchCluster(
 } // namespace mlpack
 
-#endif
+#endif
```
No need to remove the newline from the end of the file.

rcurtin reviewed (commented):
@MarkFischinger thanks for working on this! I have some comments---the strategy as it currently is unfortunately won't work, but I think we can still simplify, avoid the incorrect optimization, and merge. :+1:

MarkFischinger commented:
@rcurtin i implemented the suggestion in the search range and search range rules. Even with those changes, performance wise my benchmarks did not show a significant improvement that could not be explained by the standard deviation. The first algorithmic improvement was not aligned with the functionality of dbscan therefore i reverted it. Because of the nature of dbscan, finding a parallelization method or algorithmic improvement is not simple. I will come back to this as soon as i have a new idea.

rcurtin commented:
Yeah, the way to do it (I did it in 2015 but for some reason never opened a PR because it needed tuning that I never had the time for) is to use OpenMP tasks in the dual-tree traversers, and tune it so a new task is only spawned every handful of levels.  That was the only way I saw speedup for dual-tree algorithms.  The idea of parallel dual-tree traversers (or even single-tree traversers) is a good idea, but probably needs a more in-depth dive that maybe we could do some other time.  Realistically, DBSCAN speedup would then come just from speeding up the range search itself as that (I think) is the majority of the runtime.

If you think that we won't be able to get noticeable speedup here in this PR, then we can go ahead and close it, that's fine with me :+1: (or did I misunderstand your message and you still think there is something we could merge for improvement?  either way works for me)
